### PR TITLE
Implement Client::WriteObject().

### LINKED ISSUE #557: Create an iostream class to write to objects.
When writing (or rewriting) objects, it should be possible to use an iostream to upload the contents.

- [ ] Write the basic API.
- [ ] Write an example that performs some line-oriented upload.
- [ ] Write an example that performs some buffer-oriented upload.

### PR DESCRIPTION
**Pull request by coryan**
This fixes #557. It implements the API, a unit test, an integration
test, and the sample.

### CODE REVIEW

File: google/cloud/storage/tests/testbench.py [hunk 1]
```diff
@@ -87,7 +87,10 @@ def __init__(self, gcs_url, bucket_name, name, generation, request):
         self.object_id = bucket_name + '/o/' + name + '/' + str(generation)
         now = time.gmtime(time.time())
         timestamp = time.strftime('%Y-%m-%dT%H:%M:%SZ', now)
-        self.media = request.data
+        if request.environ.get('HTTP_TRANSFER_ENCODING', '') == 'chunked':
+            self.media = ''.join(request.environ.get('wsgi.input').readlines())
```
Comment (houglum) [inline google/cloud/storage/tests/testbench.py]: Why `''.join(foo.readlines())`?  I believe `foo.read()` should return the entire contents.

Comment (coryan) [reply google/cloud/storage/tests/testbench.py]: Fixed.  For some reason I tried that and did not work, but I was trying many things so I probably had something else wrong.

File: google/cloud/storage/examples/storage_object_samples.cc [hunk 2]
```diff
@@ -112,6 +112,35 @@ void DeleteObject(gcs::Client client, int& argc, char* argv[]) {
...
+  }
+  auto bucket_name = ConsumeArg(argc, argv);
+  auto object_name = ConsumeArg(argc, argv);
+  auto desired_size = std::stol(ConsumeArg(argc, argv));
+
+  //! [write object]
+  namespace gcs = google::cloud::storage;
+  [](gcs::Client client, std::string bucket_name, std::string object_name,
+     long desired_size) {
+    std::string const text = "Lorem ipsum dolor sit amet";
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/examples/storage_object_samples.cc]: Yup, changed to make it number of lines which is easier to implement too.

**Comment by coryan** [comment]
PTAL.

File: google/cloud/storage/examples/storage_object_samples.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (houglum) [reply google/cloud/storage/examples/storage_object_samples.cc]: Thanks!

**Pull request merged**